File: vscodeExt/206.reverse-linked-list.py
#
# @lc app=leetcode id=206 lang=python3
#
# [206] Reverse Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        pre = None
        cur = head
        while cur:
            nex = cur.next # 新建nex 是 cur 的一下位

            cur.next = pre # cur 的逆转
            pre = cur # pre 前移
            cur = nex # cur 前移
        # 当 cur 为空后，pre是第一个元素
        return pre
        
# @lc code=end

File: vscodeExt/206.reverse-linked-list.java
/*
 * @lc app=leetcode id=206 lang=java
 *
 * [206] Reverse Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode pre=null;
        ListNode cur = head;

        while (cur!=null){
            ListNode next = cur.next;
            cur.next = pre; //关键的逆序操作
            pre = cur;
            cur = next;
        }
        return pre;
        
    }
}
// @lc code=end

